ext/drx/scatter_gather_aarch64.c

```diff
@@ -38,6 +38,481 @@
+        DRX_CASE(ld1rqd, 1, OPSZ_8, false, DRX_NORMAL_FAULTING);
+#undef DRX_CASE
+
+    default: DR_ASSERT_MSG(false, "Invalid scatter/gather instruction");
+    }
+}
+
+/*
+ * Emit code to expand a scalar + vector gather load into a series of equivalent scalar
+ * loads.
+ * These instructions have memory operands of the form:
+ *     [<Xn|SP>, <Zm>.<Ts>{, <mod>}]
+ * where addresses to load/store each element are calculated by adding a base address
+ * from the scalar register Xn, to an offset read from the corresponding element of the
+ * vector index register Zm.
+ *
+ * if (is_load)
+ *     clear_inactive_elements(dst);
+ * for (e=first_active_element(); e < num_elements; e =  next_active_element()) {
+ *     if (is_load)
+ *         dst[e] = scalar_load(base, offsets[e], mod);
+ *     else
+ *         scalar_store(src[e], base, offsets[e], mod);
+ * }
+
+ * For example
+ *     ld1d   (%x0,%z26.d,lsl #3)[32byte] %p1/z -> %z27.d
+ * expands to:
+ *
+ *       dup    $0x00 lsl $0x00 -> %z27.d       ; Clear dst register
+ *       pfalse  -> %p0.b
+ *   loop:
+ *       pnext  %p1 %p0.d -> %p0.d              ; p1 = next active element of p0
+ *       b.none end                             ; if (no more active elements) goto end
+ *       lastb  %p0 %z26.d -> %x1               ; extract offset for the current element
+ *       ldr    (%x0,%x1,lsl #3)[8byte] -> %x1  ; perform the scalar load
+ *       cpy    %p0/m %x1 -> %z27.d             ; cpy loaded value to dst element
+ *       b      loop
+ *   end:
+ *       ...
+ *
+ * TODO i#5036 Add support for scatter store operations.
+ */
+static void
+expand_scalar_plus_vector(void *drcontext, instrlist_t *bb, instr_t *sg_instr,
+                          const scatter_gather_info_t *sg_info, reg_id_t scratch_gpr,
+                          reg_id_t scratch_pred, app_pc orig_app_pc)
+{
+#define EMIT(op, ...)    \
+    instrlist_preinsert( \
+        bb, sg_instr, INSTR_XL8(INSTR_CREATE_##op(drcontext, __VA_ARGS__), orig_app_pc))
+
+    DR_ASSERT_MSG(reg_is_z(sg_info->index_reg), "Index must be a Z register");
+
+    if (sg_info->is_load) {
+        /* First we deal with the inactive elements. Gather loads are always zeroing so we
+         * need to set all inactive elements to 0.
+         */
+        if (sg_info->index_reg == sg_info->gather_dst_reg) {
+            /* The dst register is also the index register so we need to preserve the
+             * value of the active elements so we can use them as offsets. We do this by
+             * cpying a 0 value into the dst register using the inverse of the mask_reg as
+             * the governing predicate.
+             */
+
+            /* ptrue    scratch_pred.b */
+            EMIT(ptrue_sve, opnd_create_reg_element_vector(scratch_pred, OPSZ_1),
+                 opnd_create_immed_pred_constr(DR_PRED_CONSTR_ALL));
+
+            /* not      scratch_pred.b, scratch_pred/z, mask_reg.b */
```

Comment:
No, I did mean `not` and `scratch_pred/z`. `not` is an alias for `eor` in AArch64 assembly, but the DynamoRIO codec doesn't support all instruction aliases so we need to encode it using the underlying `eor` encoding.

https://developer.arm.com/documentation/ddi0602/2023-06/SVE-Instructions/NOT--predicate---Bitwise-invert-predicate--an-alias-of-EOR--predicates--?lang=en

>`NOT <Pd>.B, <Pg>/Z, <Pn>.B`
is equivalent to
`EOR <Pd>.B, <Pg>/Z, <Pn>.B, <Pg>.B`

I put `not` in the comment because I think it is clearer what the intention is but I can see that if you don't happen to know that `not` is an alias for `eor`, then it does look like a mistake.

I should have just added an `INSTR_CREATE_not_sve_pred_b` macro so we can just use the alias to encode the instruction.

Reply:
> I should have just added an INSTR_CREATE_not_sve_pred_b macro so we can just use the alias to encode the instruction.

Sure we could just do that. Or a clarifying comment would work too.